### PR TITLE
Refine how we detect namespace packages

### LINKED ISSUE #12112: `pytest==8.1.1` Import regression in some namespace package layouts
As first noted in #12074, below is an example project structure with two packages under the `ns` namespace that now fail to import after the changes introduced in `8.1.x`:

```
ns
└── python
    ├── bar
    │   ├── ns
    │   │   ├── bar
    │   │   └── test
    │   │       ├── __init__.py
    │   │       ├── bar.py
    │   │       └── test_bar.py
    │   └── pyproject.toml
    └── foo
        ├── ns
        │   ├── foo
        │   └── test
        │       ├── __init__.py
        │       ├── foo.py
        │       └── test_foo.py
        └── pyproject.toml
```

Below are the contents of `test_foo.py` and `foo.py`. `test_bar.py` and `bar.py` look nearly identical.

```
# python/foo/ns/test/test_foo.py
from .foo import value

def test_foo():
    assert value == "foo"
```

```
# python/foo/ns/test/foo.py
value = "foo"
```

In `pytest==8.0.2`, `python -m pytest --import-mode=importlib` correctly discovers and runs the tests from the top level `ns` directory. In `pytest==8.1.1`, `python -m pytest --import-mode=importlib -o "consider_namespace_packages=true"`, results in the following error during collection:

```
========================== test session starts ===========================
platform darwin -- Python 3.9.16, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/user/pytest-12074
collected 0 items / 2 errors

================================= ERRORS =================================
____________ ERROR collecting python/bar/ns/test/test_bar.py _____________
ImportError while importing test module '/home/user/pytest-12074/python/bar/ns/test/test_bar.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
python/bar/ns/test/test_bar.py:1: in <module>
    from .bar import value
E   ModuleNotFoundError: No module named 'test.bar'
____________ ERROR collecting python/foo/ns/test/test_foo.py _____________
ImportError while importing test module '/home/user/pytest-12074/python/foo/ns/test/test_foo.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
python/foo/ns/test/test_foo.py:1: in <module>
    from .foo import value
E   ModuleNotFoundError: No module named 'test.foo'
======================== short test summary info =========================
ERROR python/bar/ns/test/test_bar.py
ERROR python/foo/ns/test/test_foo.py
!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!
=========================== 2 errors in 0.04s ============================
```

To reproduce the issue clone https://github.com/aaraney/pytest-12074, pip install `ns.foo` and `ns.bar`, pip install the different versions of `pytest` and run with the aforementioned commands.

### PR DESCRIPTION
**Pull request by nicoddemus**
Previously we used a hand crafted approach to detect namespace packages, however we should rely on ``importlib`` to detect them for us.

Fix #12112

### CODE REVIEW

File: src/_pytest/pathlib.py [hunk 1]
```diff
@@ -795,6 +787,35 @@ def resolve_pkg_root_and_module_name(
...
+    # If the path has na __init__.py file, it means it is not
+    # a namespace package:.
+    # https://packaging.python.org/en/latest/guides/packaging-namespace-packages.
+    if (module_path / "__init__.py").is_file():
+        return False
+
+    module_name = module_path.name
+
+    # Empty module names break find_spec.
+    if not module_name:
```
Comment (bluetech) [inline src/_pytest/pathlib.py:800]: I'm curious, what `Path` has empty `name`?

File: src/_pytest/pathlib.py [hunk 2]
```diff
@@ -795,6 +787,35 @@ def resolve_pkg_root_and_module_name(
...
+        return False
+
+    # Modules starting with "." indicate relative imports and break find_spec, and we are only attempting
+    # to find top-level namespace packages anyway.
+    if module_name.startswith("."):
```
Comment (bluetech) [inline src/_pytest/pathlib.py:805]: What if the dir contains a `.` not at the beginning e.g. `dir/py.test`?

File: src/_pytest/pathlib.py [hunk 3]
```diff
@@ -795,6 +787,35 @@ def resolve_pkg_root_and_module_name(
     raise CouldNotResolvePathError(f"Could not resolve for {path}")
 
 
+def _is_namespace_package(module_path: Path) -> bool:
+    # If the path has na __init__.py file, it means it is not
+    # a namespace package:.
+    # https://packaging.python.org/en/latest/guides/packaging-namespace-packages.
```
Comment (bluetech) [inline src/_pytest/pathlib.py:793]:
```suggestion
    # If the path has an __init__.py file, it means it is not
    # a namespace package:
    # https://packaging.python.org/en/latest/guides/packaging-namespace-packages
```

File: src/_pytest/pathlib.py [hunk 4]
```diff
@@ -795,6 +787,35 @@ def resolve_pkg_root_and_module_name(
     raise CouldNotResolvePathError(f"Could not resolve for {path}")
 
 
+def _is_namespace_package(module_path: Path) -> bool:
+    # If the path has na __init__.py file, it means it is not
```
Comment (bluetech) [inline src/_pytest/pathlib.py:791]: Previous comment said

> If any of the parent paths has a __init__.py, it means it is not a namespace package (see the docs linked above).

the difference being "any", i.e. previously terminated the entire loop, now only returns `False` for the specific level but keeps trying upwards. Which is correct?

**Review by bluetech — COMMENTED** [review]
The new approach makes sense to me (as far as I understand the details). I left a couple of questions.

File: src/_pytest/pathlib.py (hunk 1 above)
Comment (nicoddemus) [reply src/_pytest/pathlib.py:800]: Surprised me at first too, but:

```
>>> import pathlib
>>> pathlib.Path(".").name
''
```

File: src/_pytest/pathlib.py (hunk 3 above)
Comment (nicoddemus) [reply src/_pytest/pathlib.py:793]: Thanks!

File: src/_pytest/pathlib.py (hunk 4 above)
Comment (nicoddemus) [reply src/_pytest/pathlib.py]: Good catch, will check on this.

File: src/_pytest/pathlib.py (hunk 2 above)
Comment (nicoddemus) [reply src/_pytest/pathlib.py:805]: We can drop this check and catch an error from `find_spec` instead. 👍

**Comment by nicoddemus** [comment]
I devised another corner case which is not covered yet; I pushed a test, but won't have time to work on it anymore today, I will pick it up later.

**Comment by nicoddemus** [comment]
Note: also test modules are not being imported more than once.

File: changelog/12112.bugfix.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Improve namespace packages detection when :confval:`consider_namespace_packages` is enabled, covering more situations (like editable installs).
```
[the remark this answers is not in the document]
Comment (nicoddemus) [reply changelog/12112.bugfix.rst:1]: Given the changes here, butter make this an improvement.

File: src/_pytest/pathlib.py [hunk 6]
```diff
@@ -762,39 +775,63 @@ def resolve_pkg_root_and_module_name(
...
+    if pkg_root is not None:
+        module_name = compute_module_name(pkg_root, path)
         return pkg_root, module_name
 
     raise CouldNotResolvePathError(f"Could not resolve for {path}")
 
 
+def is_importable(root: Path, module_path: Path) -> bool:
+    """
+    Return if the given module path could be imported normally by Python, akin to the user
```
Comment (bluetech) [inline src/_pytest/pathlib.py:805]: It will be helpful to explain the `root` parameter

File: src/_pytest/pathlib.py [hunk 7]
```diff
@@ -762,39 +775,63 @@ def resolve_pkg_root_and_module_name(
...
+    entering the REPL and importing the corresponding module name directly.
+    """
+    module_name = compute_module_name(root, module_path)
+    try:
+        # Note this is different from what we do in ``import_path``, where we also search sys.meta_path.
```
Comment (bluetech) [inline src/_pytest/pathlib.py:810]: I'm a bit confused - doesn't `importlib.util.find_spec` itself search in `meta_path`?

File: src/_pytest/pathlib.py (hunk 6 above)
Comment (nicoddemus) [reply src/_pytest/pathlib.py:805]: Did that, but end up realizing that `module_name` here makes more sense and easier to understand. Please take a look.

File: src/_pytest/pathlib.py (hunk 7 above)
Comment (nicoddemus) [reply src/_pytest/pathlib.py:810]: It does, the wording was a big off. Made another pass at it, please take a look. 👍

**Review by bluetech — APPROVED** [review]
Looks good to me!

**Comment by nicoddemus** [comment]
Thanks @bluetech for the review.

I will leave this open for a few more days to give others a chance to review too (which would be appreciated).

**Review by RonnyPfannschmidt — APPROVED** [review]

**Comment by flying-sheep** [comment]
Hi, this got marked as an improvement. Isn’t this a bugfix?

I’m asking because we’re trying to gauge how long we can expect to wait for a release containing this.

**Comment by nicoddemus** [comment]
Good question, decided to go with "improvement" because it contains significant changes on how things work -- seemed safe to release this in the next minor release instead of a bug-fix release.

But it is not set in stone, if we think this is OK to go as a bug fix we can change that, for sure.
